### PR TITLE
Add Vec128 sort

### PR DESCRIPTION
**Pull request by TheNumbat**
1. Deletes the specifc vec128 type info from the lambda vec128 layout.
    - This information was unused, and seemed non-trivial to preserve via the frontend layouts. It also wasn't really useful in the first place, since all simd operations define the relevant vector layouts.
3. Adds a new `Vec128` sort and unboxed versions of our 128-bit simd vector types.
4. Duplicates the simd vector tests for unboxed vectors.
5. Adds box/unbox primitives for vec128 layout

Reviewable by commit.

Three features remain unsupported:
- Fields of unboxed vector type (#3077)
- Arrays of unboxed vectors (#2969)
- String/bytes/bigstring load/store for unboxed vectors (#2966)

Currently the compiler will raise in these cases, but they will be addressed in later PRs.

### CODE REVIEW

File: tests/simd/basic_u.ml
Comment (ccasin) [inline tests/simd/basic_u.ml:2]: You've duplicated the tests for boxed simd, which is great, but we're also going to need a bunch more tests: typing tests for the new type and sort, and possibly some more runtime tests.

You will notice for each sort we have a test directory `ocaml/testsuite/tests/typing-layouts-<sort>`.  Please look, for example, at the `bits64` tests and duplicate them for the new sort.  The runtime tests in `unboxed_bits64s.ml` are somewhat duplicative of the runtime tests you have here, and may not need to be duplicated (but check you've got all those cases here).  The typing tests should ~all be duplicated, and consider if any new typing tests should be added for the new types.  (E.g., add a test showing if you try to use `int64x2#` where `int32x4#` is expected, you get a type error).

**Review by ccasin — COMMENTED** [review]
Really sorry that, due to ICFP prep, ICFP, and the sprint, it took me so long to get to this.  Various comments below.  I'd also like to talk through the story about how this will work in the tree (want to avoid jsoo issues like we had with float32), but perhaps we can talk through that when we meet tomorrow.

I've read everything under `ocaml`.  Please also find appropriate reviewers for the middleend and backend bits.  (It may make sense for them to review the tests as well, but probably you'll want to address my requests for more tests first)

File: ocaml/lambda/lambda.ml
Comment (TheNumbat) [reply ocaml/lambda/lambda.ml:452]: Yeah, it had no purpose and propagating the information via the vec128 layout would have required some larger changes. This information could theoretically be useful for optimizations in flambda2, but all operations on vec128s imply a certain type anyway.

File: ocaml/typing/typeopt.ml
Comment (TheNumbat) [reply ocaml/typing/typeopt.ml]: Added a beta maturity to simd as well as layouts

File: ocaml/typing/typeopt.ml
Comment (TheNumbat) [reply ocaml/typing/typeopt.ml:201]: Will be fixed in a following PR

File: ocaml/typing/typedecl.ml
Comment (TheNumbat) [reply ocaml/typing/typedecl.ml]: Will be fixed in a following PR

File: tests/simd/basic_u.ml
Comment (TheNumbat) [reply tests/simd/basic_u.ml:1]: Added comments/crs

File: ocaml/typing/jkind.ml
Comment (TheNumbat) [reply ocaml/typing/jkind.ml]: Moved to beta maturity

File: ocaml/typing/predef.ml
Comment (TheNumbat) [reply ocaml/typing/predef.ml:569]: Moved to beta maturity

File: tests/simd/basic_u.ml
Comment (TheNumbat) [reply tests/simd/basic_u.ml:2]: I've duplicated the typing tests under `typing-layouts-vec128`, but the runtime tests (`test_int64_u.ml`/`alloc.ml`/`unboxed_bits64.ml`) can't be directly ported, and I think these are covered by the unboxed SIMD tests under `tests/simd`.

Comment (TheNumbat) [reply tests/simd/basic_u.ml:2]: Also note the new tests include fatal errors when using vec128 fields; this will be fixed in a following PR

**Comment by TheNumbat** [comment]
@mshinwell has looked over the changes under `middle_end/flambda2`

**Comment by mshinwell** [comment]
Not really, I just had a question!

**Review by ccasin — APPROVED** [review]
Approved for the front-end bits.  Obviously don't merge until we're also ready to merge the PRs fixing the fatal errors.

File: ocaml/testsuite/tests/typing-layouts-vec128/basics_beta.ml
Comment (TheNumbat) [reply ocaml/testsuite/tests/typing-layouts-vec128/basics_beta.ml:2]: Currently the only difference is that alpha enables `or_null`, so the error messages print `value_or_null` instead of `value` in several places. Is that worth keeping?

**Review by mshinwell — APPROVED** [review]

File: ocaml/testsuite/tests/typing-layouts-vec128/basics_beta.ml
Comment (ccasin) [reply ocaml/testsuite/tests/typing-layouts-vec128/basics_beta.ml:2]: Aha, that's fine, we can combine them after that is changed

**Pull request merged**
